vyper/semantics/types/function.py

```diff
@@ -507,7 +530,10 @@ def implements(self, other: "ContractFunctionT") -> bool:
         if return_type and not return_type.compare_type(other_return_type):  # type: ignore
             return False
 
-        return self.mutability == other.mutability
+        if self.mutability != other.mutability:
+            return False
+
+        return self.visibility == other.visibility
```

Comment:
i think returning False here is unreachable, not sure why this change is needed

Reply:
currently this should be true. added the case for the sake of future-proofing the func in case the caller's semantics change / new callers are added

Reply:
hmm. in that case we should have an assertion, to communicate to the reader what the assumption is, not leave the reader combing the codebase trying to find cases where the two visibilities are different.

Reply:
addressed in: https://github.com/vyperlang/vyper/pull/4178/commits/1e3541d83b8d0f70410751cec7bdf3ac51dac6a1